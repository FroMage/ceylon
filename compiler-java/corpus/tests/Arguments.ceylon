class Arguments() {
	
	class NoArguments() {
		
		void greet(String value="Hello") {
			log.info(value);
		}
		
		greet();
		greet {};
		
		class Greeting(String value="Hello") {
			log.info(value);
		}
		
		Greeting();
		Greeting {};
		
	}
	
	class SingleArgument() {
		
		void outp(String value) {
			log.info(value);
		}
		
		outp("Hello");
		outp { value="Hi"; };
		
		class Output(String value) {
			log.info(value);
		}
		
		Output("Hello");
		Output { value="Hi"; };
		
	}
	
	class MultipleArguments() {
	
		String fullName(String firstName, String lastName) {
			return firstName + " " + lastName;
		}
		
		String methodResultWithPositionalParams = fullName("Gavin", "King");
		String methodResultWithNamedParams = fullName { firstName="Andrew"; lastName="Haley"; };
		
		class FullName(String firstName, String lastName) {
			String value = firstName + " " + lastName;
		}
		
		FullName instantiationResultWithPositionalParams = FullName("Gavin", "King");
		FullName instantiationResultWithNamedParams = FullName { firstName="Andrew"; lastName="Haley"; };
		
	}
	
	class Varargs() {
	
		String join(String sep, String... strings) {
			return (sep+" ").join(strings);
		}
		
		String methodResultWithPositionalVarargs = join(",", "foo", "bar"); 
		String methodResultWithNamedVarargs = join { sep=","; "one", "two" };
		
		String methodResultWithPositionalEnumerationParam = join( ",", {"foo", "bar"} ); 
		String methodResultWithNamedEnumerationParam = join { sep=","; strings={"one", "two"}; };
		
		List<String> instantiationResultWithPositionalVarargs = List("foo", "bar", "baz");
		List<String> instantiationResultWithNamedVarargs = List { "one", "two", "three" };
		
		List<String> instantiationResultWithPositionalEnumerationParam = List( {"foo", "bar", "baz"} );
		List<String> instantiationResultWithNamedEnumerationParam = List { strings={"one", "two", "three"}; };
		
	}
	
	class FunctionalArguments() {
	
		void logLazily(String message()) {
			if (log.infoEnabled) { 
				log.info( message() );
			}
		}
		
		logLazily() message ("hello");
		logLazily() message { return "hello" };
		logLazily { String message() { return "hello" } };
	
	
		public void from<Y>(Y initial, 
							Boolean until(Y y), 
							Y each(Y y), 
							void perform(Y y)) {
			while (!until(y)) {
				perform(y);
				y:=each(y);
			}
		}
		
		from(0) 
			until (Natural y) (y==10)
			each (Natural y) (y+2)
			perform (Natural y) { log.info(y); };
		
		from(0) 
			until(Natural y) { return y==10 } 
			each(Natural y) { return y+2 } 
			perform(Natural y) { log.info(y); };
		
		from {
			initial=0; 
			Boolean until(Natural y) { return y==10 } 
			Natural each(Natural y) { return y+2 } 
			void perform(Natural y) { log.info(y); }
		};
		
		class Processor<X,Y>(Y process(X x)) {
			Y handle(X x) { return process(x).strip }
		}
		
		Processor<Float,String> p = 
			Processor() process (Float f) ($f);
		
		Processor<Float,String> q = 
			Processor() process (Float f) { return $f };
		
		Processor<Float,String> r = 
			Processor { 
				String process (Float f) { return $f }
			};
			
	}

}