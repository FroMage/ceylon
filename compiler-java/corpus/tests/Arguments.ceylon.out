# Arguments.ceylon

(TYPE_DECL
  (TYPE_DECL
    (CLASS_DECL
      (TYPE_NAME Arguments)
      (STMT_LIST
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME NoArguments)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME greet)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME value)
                    (INIT_EXPR
                      (EXPR
                        (STRING_CST "Hello")))))
                (STMT_LIST
                  (EXPR
                    (EXPR
                      (MEMBER_NAME log) .
                      (MEMBER_NAME info)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (MEMBER_NAME value))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME greet)
                  (CALL_EXPR ARG_LIST)))
              (EXPR
                (EXPR
                  (MEMBER_NAME greet)
                  (CALL_EXPR ARG_LIST)))
              (TYPE_DECL
                (CLASS_DECL
                  (TYPE_NAME Greeting)
                  (FORMAL_PARAMETER_LIST
                    (FORMAL_PARAMETER
                      (TYPE
                        (TYPE_NAME String))
                      (ARG_NAME value)
                      (INIT_EXPR
                        (EXPR
                          (STRING_CST "Hello")))))
                  (STMT_LIST
                    (EXPR
                      (EXPR
                        (MEMBER_NAME log) .
                        (MEMBER_NAME info)
                        (CALL_EXPR
                          (ARG_LIST
                            (EXPR
                              (MEMBER_NAME value)))))))))
              (EXPR
                (EXPR
                  (TYPE_NAME Greeting)
                  (CALL_EXPR ARG_LIST)))
              (EXPR
                (EXPR
                  (TYPE_NAME Greeting)
                  (CALL_EXPR ARG_LIST))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME SingleArgument)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME outp)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME value)))
                (STMT_LIST
                  (EXPR
                    (EXPR
                      (MEMBER_NAME log) .
                      (MEMBER_NAME info)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (MEMBER_NAME value))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME outp)
                  (CALL_EXPR
                    (ARG_LIST
                      (EXPR
                        (STRING_CST "Hello"))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME outp)
                  (CALL_EXPR
                    (ARG_LIST
                      (NAMED_ARG
                        (ARG_NAME value)
                        (INIT_EXPR
                          (EXPR
                            (STRING_CST "Hi"))))))))
              (TYPE_DECL
                (CLASS_DECL
                  (TYPE_NAME Output)
                  (FORMAL_PARAMETER_LIST
                    (FORMAL_PARAMETER
                      (TYPE
                        (TYPE_NAME String))
                      (ARG_NAME value)))
                  (STMT_LIST
                    (EXPR
                      (EXPR
                        (MEMBER_NAME log) .
                        (MEMBER_NAME info)
                        (CALL_EXPR
                          (ARG_LIST
                            (EXPR
                              (MEMBER_NAME value)))))))))
              (EXPR
                (EXPR
                  (TYPE_NAME Output)
                  (CALL_EXPR
                    (ARG_LIST
                      (EXPR
                        (STRING_CST "Hello"))))))
              (EXPR
                (EXPR
                  (TYPE_NAME Output)
                  (CALL_EXPR
                    (ARG_LIST
                      (NAMED_ARG
                        (ARG_NAME value)
                        (INIT_EXPR
                          (EXPR
                            (STRING_CST "Hi")))))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME MultipleArguments)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME fullName)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME firstName))
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME lastName)))
                (STMT_LIST
                  (RET_STMT
                    (EXPR
                      (+
                        (+
                          (MEMBER_NAME firstName)
                          (STRING_CST " "))
                        (MEMBER_NAME lastName))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithPositionalParams)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME fullName)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (STRING_CST "Gavin"))
                          (EXPR
                            (STRING_CST "King"))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithNamedParams)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME fullName)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (ARG_NAME firstName)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST "Andrew"))))
                          (NAMED_ARG
                            (ARG_NAME lastName)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST "Haley"))))))))))
              (TYPE_DECL
                (CLASS_DECL
                  (TYPE_NAME FullName)
                  (FORMAL_PARAMETER_LIST
                    (FORMAL_PARAMETER
                      (TYPE
                        (TYPE_NAME String))
                      (ARG_NAME firstName))
                    (FORMAL_PARAMETER
                      (TYPE
                        (TYPE_NAME String))
                      (ARG_NAME lastName)))
                  (STMT_LIST
                    (MEMBER_DECL
                      (MEMBER_TYPE
                        (TYPE
                          (TYPE_NAME String)))
                      (MEMBER_NAME value)
                      (INIT_EXPR
                        (EXPR
                          (+
                            (+
                              (MEMBER_NAME firstName)
                              (STRING_CST " "))
                            (MEMBER_NAME lastName))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME FullName)))
                (MEMBER_NAME instantiationResultWithPositionalParams)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME FullName)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (STRING_CST "Gavin"))
                          (EXPR
                            (STRING_CST "King"))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME FullName)))
                (MEMBER_NAME instantiationResultWithNamedParams)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME FullName)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (ARG_NAME firstName)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST "Andrew"))))
                          (NAMED_ARG
                            (ARG_NAME lastName)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST "Haley")))))))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME Varargs)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME join)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME sep))
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String)) ...
                    (ARG_NAME strings)))
                (STMT_LIST
                  (RET_STMT
                    (EXPR
                      (EXPR
                        (EXPR
                          (+
                            (MEMBER_NAME sep)
                            (STRING_CST " "))) .
                        (MEMBER_NAME join)
                        (CALL_EXPR
                          (ARG_LIST
                            (EXPR
                              (MEMBER_NAME strings)))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithPositionalVarargs)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME join)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (STRING_CST ","))
                          (EXPR
                            (STRING_CST "foo"))
                          (EXPR
                            (STRING_CST "bar"))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithNamedVarargs)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME join)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (ARG_NAME sep)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST ","))))
                          (VARARGS
                            (EXPR_LIST
                              (EXPR
                                (STRING_CST "one"))
                              (EXPR
                                (STRING_CST "two"))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithPositionalEnumerationParam)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME join)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (STRING_CST ","))
                          (EXPR
                            (ENUM_LIST
                              (EXPR_LIST
                                (EXPR
                                  (STRING_CST "foo"))
                                (EXPR
                                  (STRING_CST "bar")))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME methodResultWithNamedEnumerationParam)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (MEMBER_NAME join)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (ARG_NAME sep)
                            (INIT_EXPR
                              (EXPR
                                (STRING_CST ","))))
                          (NAMED_ARG
                            (ARG_NAME strings)
                            (INIT_EXPR
                              (EXPR
                                (ENUM_LIST
                                  (EXPR_LIST
                                    (EXPR
                                      (STRING_CST "one"))
                                    (EXPR
                                      (STRING_CST "two")))))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME List)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME instantiationResultWithPositionalVarargs)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME List)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (STRING_CST "foo"))
                          (EXPR
                            (STRING_CST "bar"))
                          (EXPR
                            (STRING_CST "baz"))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME List)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME instantiationResultWithNamedVarargs)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME List)
                      (CALL_EXPR
                        (ARG_LIST
                          (VARARGS
                            (EXPR_LIST
                              (EXPR
                                (STRING_CST "one"))
                              (EXPR
                                (STRING_CST "two"))
                              (EXPR
                                (STRING_CST "three"))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME List)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME instantiationResultWithPositionalEnumerationParam)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME List)
                      (CALL_EXPR
                        (ARG_LIST
                          (EXPR
                            (ENUM_LIST
                              (EXPR_LIST
                                (EXPR
                                  (STRING_CST "foo"))
                                (EXPR
                                  (STRING_CST "bar"))
                                (EXPR
                                  (STRING_CST "baz")))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME List)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME instantiationResultWithNamedEnumerationParam)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME List)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (ARG_NAME strings)
                            (INIT_EXPR
                              (EXPR
                                (ENUM_LIST
                                  (EXPR_LIST
                                    (EXPR
                                      (STRING_CST "one"))
                                    (EXPR
                                      (STRING_CST "two"))
                                    (EXPR
                                      (STRING_CST "three"))))))))))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME FunctionalArguments)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME logLazily)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME message)))
                (STMT_LIST
                  (IF_STMT
                    (CONDITION
                      (EXPR
                        (EXPR
                          (MEMBER_NAME log) .
                          (MEMBER_NAME infoEnabled))))
                    (IF_TRUE
                      (STMT_LIST
                        (EXPR
                          (EXPR
                            (MEMBER_NAME log) .
                            (MEMBER_NAME info)
                            (CALL_EXPR
                              (ARG_LIST
                                (EXPR
                                  (EXPR
                                    (MEMBER_NAME message)
                                    (CALL_EXPR ARG_LIST))))))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME logLazily)
                  (CALL_EXPR ARG_LIST
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME message))
                      (ANON_METH
                        (EXPR
                          (STRING_CST "hello")))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME logLazily)
                  (CALL_EXPR ARG_LIST
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME message))
                      (ANON_METH
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (STRING_CST "hello")))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME logLazily)
                  (CALL_EXPR
                    (ARG_LIST
                      (NAMED_ARG
                        (TYPE
                          (TYPE_NAME String))
                        (ARG_NAME message)
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (STRING_CST "hello")))))))))
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME from)
                (TYPE_PARAMETER_LIST
                  (TYPE_PARAMETER
                    (TYPE_NAME Y)))
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME Y))
                    (ARG_NAME initial))
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME Boolean))
                    (ARG_NAME until)
                    (FORMAL_PARAMETER_LIST
                      (FORMAL_PARAMETER
                        (TYPE
                          (TYPE_NAME Y))
                        (ARG_NAME y))))
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME Y))
                    (ARG_NAME each)
                    (FORMAL_PARAMETER_LIST
                      (FORMAL_PARAMETER
                        (TYPE
                          (TYPE_NAME Y))
                        (ARG_NAME y))))
                  (FORMAL_PARAMETER void
                    (ARG_NAME perform)
                    (FORMAL_PARAMETER_LIST
                      (FORMAL_PARAMETER
                        (TYPE
                          (TYPE_NAME Y))
                        (ARG_NAME y)))))
                (STMT_LIST
                  (WHILE_STMT
                    (CONDITION
                      (EXPR
                        (!
                          (EXPR
                            (MEMBER_NAME until)
                            (CALL_EXPR
                              (ARG_LIST
                                (EXPR
                                  (MEMBER_NAME y))))))))
                    (WHILE_BLOCK
                      (STMT_LIST
                        (EXPR
                          (EXPR
                            (MEMBER_NAME perform)
                            (CALL_EXPR
                              (ARG_LIST
                                (EXPR
                                  (MEMBER_NAME y))))))
                        (EXPR
                          (:=
                            (MEMBER_NAME y)
                            (EXPR
                              (EXPR
                                (MEMBER_NAME each)
                                (CALL_EXPR
                                  (ARG_LIST
                                    (EXPR
                                      (MEMBER_NAME y))))))))))))
                (ANNOTATION_LIST
                  (ANNOTATION public)))
              (EXPR
                (EXPR
                  (MEMBER_NAME from)
                  (CALL_EXPR
                    (ARG_LIST
                      (EXPR
                        (INT_CST 0)))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME until))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (EXPR
                          (==
                            (MEMBER_NAME y)
                            (INT_CST 10)))))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME each))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (EXPR
                          (+
                            (MEMBER_NAME y)
                            (INT_CST 2)))))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME perform))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (EXPR
                            (EXPR
                              (MEMBER_NAME log) .
                              (MEMBER_NAME info)
                              (CALL_EXPR
                                (ARG_LIST
                                  (EXPR
                                    (MEMBER_NAME y))))))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME from)
                  (CALL_EXPR
                    (ARG_LIST
                      (EXPR
                        (INT_CST 0)))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME until))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (==
                                (MEMBER_NAME y)
                                (INT_CST 10)))))))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME each))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (+
                                (MEMBER_NAME y)
                                (INT_CST 2)))))))
                    (NAMED_ARG
                      (ARG_NAME
                        (ARG_NAME perform))
                      (ANON_METH
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (EXPR
                            (EXPR
                              (MEMBER_NAME log) .
                              (MEMBER_NAME info)
                              (CALL_EXPR
                                (ARG_LIST
                                  (EXPR
                                    (MEMBER_NAME y))))))))))))
              (EXPR
                (EXPR
                  (MEMBER_NAME from)
                  (CALL_EXPR
                    (ARG_LIST
                      (NAMED_ARG
                        (ARG_NAME initial)
                        (INIT_EXPR
                          (EXPR
                            (INT_CST 0))))
                      (NAMED_ARG
                        (TYPE
                          (TYPE_NAME Boolean))
                        (ARG_NAME until)
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (==
                                (MEMBER_NAME y)
                                (INT_CST 10))))))
                      (NAMED_ARG
                        (TYPE
                          (TYPE_NAME Natural))
                        (ARG_NAME each)
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (RET_STMT
                            (EXPR
                              (+
                                (MEMBER_NAME y)
                                (INT_CST 2))))))
                      (NAMED_ARG void
                        (ARG_NAME perform)
                        (FORMAL_PARAMETER_LIST
                          (FORMAL_PARAMETER
                            (TYPE
                              (TYPE_NAME Natural))
                            (ARG_NAME y)))
                        (STMT_LIST
                          (EXPR
                            (EXPR
                              (MEMBER_NAME log) .
                              (MEMBER_NAME info)
                              (CALL_EXPR
                                (ARG_LIST
                                  (EXPR
                                    (MEMBER_NAME y))))))))))))
              (TYPE_DECL
                (CLASS_DECL
                  (TYPE_NAME Processor)
                  (TYPE_PARAMETER_LIST
                    (TYPE_PARAMETER
                      (TYPE_NAME X))
                    (TYPE_PARAMETER
                      (TYPE_NAME Y)))
                  (FORMAL_PARAMETER_LIST
                    (FORMAL_PARAMETER
                      (TYPE
                        (TYPE_NAME Y))
                      (ARG_NAME process)
                      (FORMAL_PARAMETER_LIST
                        (FORMAL_PARAMETER
                          (TYPE
                            (TYPE_NAME X))
                          (ARG_NAME x)))))
                  (STMT_LIST
                    (MEMBER_DECL
                      (MEMBER_TYPE
                        (TYPE
                          (TYPE_NAME Y)))
                      (MEMBER_NAME handle)
                      (FORMAL_PARAMETER_LIST
                        (FORMAL_PARAMETER
                          (TYPE
                            (TYPE_NAME X))
                          (ARG_NAME x)))
                      (STMT_LIST
                        (RET_STMT
                          (EXPR
                            (EXPR
                              (MEMBER_NAME process)
                              (CALL_EXPR
                                (ARG_LIST
                                  (EXPR
                                    (MEMBER_NAME x)))) .
                              (MEMBER_NAME strip)))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Processor)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME Float))
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME p)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME Processor)
                      (CALL_EXPR ARG_LIST
                        (NAMED_ARG
                          (ARG_NAME
                            (ARG_NAME process))
                          (ANON_METH
                            (FORMAL_PARAMETER_LIST
                              (FORMAL_PARAMETER
                                (TYPE
                                  (TYPE_NAME Float))
                                (ARG_NAME f)))
                            (EXPR
                              (PREFIX_EXPR $
                                (MEMBER_NAME f))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Processor)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME Float))
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME q)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME Processor)
                      (CALL_EXPR ARG_LIST
                        (NAMED_ARG
                          (ARG_NAME
                            (ARG_NAME process))
                          (ANON_METH
                            (FORMAL_PARAMETER_LIST
                              (FORMAL_PARAMETER
                                (TYPE
                                  (TYPE_NAME Float))
                                (ARG_NAME f)))
                            (STMT_LIST
                              (RET_STMT
                                (EXPR
                                  (PREFIX_EXPR $
                                    (MEMBER_NAME f))))))))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Processor)
                    (TYPE_ARG_LIST
                      (TYPE
                        (TYPE_NAME Float))
                      (TYPE
                        (TYPE_NAME String)))))
                (MEMBER_NAME r)
                (INIT_EXPR
                  (EXPR
                    (EXPR
                      (TYPE_NAME Processor)
                      (CALL_EXPR
                        (ARG_LIST
                          (NAMED_ARG
                            (TYPE
                              (TYPE_NAME String))
                            (ARG_NAME process)
                            (FORMAL_PARAMETER_LIST
                              (FORMAL_PARAMETER
                                (TYPE
                                  (TYPE_NAME Float))
                                (ARG_NAME f)))
                            (STMT_LIST
                              (RET_STMT
                                (EXPR
                                  (PREFIX_EXPR $
                                    (MEMBER_NAME f)))))))))))))))))))
